File: scripts/build_schlicht_seed.php
#!/usr/bin/env php
<?php
/**
 * Baut seed/layout-sets/schlicht/ aus den Standard-Folienvorlagen in seed/templates/.
 * Reihenfolge und Inhalte entsprechen dem Folien-Set „Schlicht“.
 *
 * Usage: php scripts/build_schlicht_seed.php
 */
require __DIR__ . '/../config.php';

$seedName = 'schlicht';
$target = BASE_PATH . '/seed/layout-sets/' . $seedName;
$templateOrder = [
    '77da9b45b679c85b', // Titel
    '6f46aa3644fc25d5', // Abschnitt
    '60c1046a44ddcd8f', // Überschrift
    '4a38cb88d0722919', // Überschrift und Inhalt
    '1040b5e23d947b36', // Überschrift und zwei Inhalte
    '0691915389a265ac', // Text
    'fc35b144beff7c31', // Vergleich
];

if (is_dir($target)) {
    $it = new RecursiveIteratorIterator(
        new RecursiveDirectoryIterator($target, RecursiveDirectoryIterator::SKIP_DOTS),
        RecursiveIteratorIterator::CHILD_FIRST
    );
    foreach ($it as $f) {
        $f->isDir() ? rmdir($f->getPathname()) : unlink($f->getPathname());
    }
    rmdir($target);
}
mkdir($target, 0770, true);
mkdir($target . '/assets', 0770, true);

$slides = [];
foreach ($templateOrder as $templateSeedId) {
    $src = BASE_PATH . '/seed/templates/' . $templateSeedId;
    $metaPath = $src . '/meta.json';
    $slidesPath = $src . '/slides.json';
    if (!is_file($metaPath) || !is_file($slidesPath)) {
        fwrite(STDERR, "Fehlende Vorlage: {$templateSeedId}\n");
        exit(1);
    }
    $tplMeta = json_decode((string)file_get_contents($metaPath), true);
    $tplSlides = json_decode((string)file_get_contents($slidesPath), true);
    $slide = $tplSlides['slides'][0] ?? null;
    if (!is_array($slide)) {
        fwrite(STDERR, "Leere Vorlage: {$templateSeedId}\n");
        exit(1);
    }

    $title = trim((string)($tplMeta['title'] ?? ''));
    $slide['id'] = Storage::generateId(4);
    $slide['layoutKey'] = LayoutSet::layoutKeyFromTitle($title);
    $slide['label'] = $title !== '' ? $title : LayoutSet::slideLabel($slide);
    $slide = LayoutSet::prepareLayoutSlide($slide);

    $json = json_encode($slide, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    $json = str_replace(
        'asset.php?id=' . urlencode($templateSeedId) . '&',
        'asset.php?id=' . urlencode($seedName) . '&',
        $json
    );
    $slide = json_decode($json, true) ?? $slide;
    $slides[] = $slide;

    $srcAssets = $src . '/assets';
    if (is_dir($srcAssets)) {
        foreach (glob($srcAssets . '/*') ?: [] as $file) {
            if (is_file($file)) {
                copy($file, $target . '/assets/' . basename($file));
            }
        }
    }
}

$seedMeta = [
    'title' => 'Schlicht',
    'width' => DEFAULT_SLIDE_WIDTH,
    'height' => DEFAULT_SLIDE_HEIGHT,
    'template_order' => 0,
    'is_layout_set' => true,
    'template_shared' => true,
    'default_layout_set' => true,
    'logosNotesOrder' => LayoutSet::DEFAULT_NOTES_ORDER,
    'elementZones' => LayoutSet::DEFAULT_ELEMENT_ZONES,
    'safe_margin' => 100,
];

file_put_contents(
    $target . '/meta.json',
    json_encode($seedMeta, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES) . "\n"
);
file_put_contents(
    $target . '/slides.json',
    json_encode(['slides' => $slides], JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES) . "\n"
);

echo "Erstellt: seed/layout-sets/{$seedName}/ mit " . count($slides) . " Layout-Folien.\n";

File: scripts/build_schlicht_seed.py
#!/usr/bin/env python3
"""Erstellt seed/layout-sets/schlicht/ aus seed/templates/ (ohne PHP)."""
import json
import re
import shutil
import secrets
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SEED_NAME = "schlicht"
TARGET = ROOT / "seed" / "layout-sets" / SEED_NAME
TEMPLATE_ORDER = [
    "77da9b45b679c85b",
    "6f46aa3644fc25d5",
    "60c1046a44ddcd8f",
    "4a38cb88d0722919",
    "1040b5e23d947b36",
    "0691915389a265ac",
    "fc35b144beff7c31",
]

ROLE_LABELS = {
    "document_title": "Titel",
    "subtitle": "Untertitel",
    "heading1": "Überschrift 1",
    "heading2": "Überschrift 2",
    "heading3": "Überschrift 3",
    "heading4": "Überschrift 4",
    "heading5": "Überschrift 5",
}


def layout_key(title: str) -> str:
    key = re.sub(r"[^a-z0-9]+", "_", title.lower().strip())
    return key.strip("_") or "layout"


def normalize(text: str) -> str:
    text = text.strip().strip("«»\"'")
    text = re.sub(r"\s+", " ", text)
    return text.lower()


def infer_role(text: str):
    norm = normalize(text)
    if norm in ("titel", "title"):
        return "document_title"
    if norm in ("untertitel", "subtitle"):
        return "subtitle"
    for role, label in ROLE_LABELS.items():
        if norm == label.lower():
            return role
    m = re.match(r"^überschrift\s*(\d)$", norm)
    if m:
        return f"heading{m.group(1)}"
    return None


def prepare_slide(slide: dict, title: str) -> dict:
    slide = json.loads(json.dumps(slide))
    slide["id"] = secrets.token_hex(2)
    slide["layoutKey"] = layout_key(title)
    slide["label"] = title
    assigned = {}
    objects = []
    for obj in slide.get("objects", []):
        if obj.get("type") != "text":
            objects.append(obj)
            continue
        role = obj.get("setRole") or obj.get("logosRole") or ""
        if not role:
            role = infer_role(obj.get("text", "")) or ""
        if role and role not in assigned:
            obj = dict(obj)
            obj["setRole"] = role
            obj.pop("logosRole", None)
            assigned[role] = True
            objects.append(obj)
        elif role:
            continue
        else:
            objects.append(obj)
    slide["objects"] = objects
    return slide


def main():
    if TARGET.exists():
        shutil.rmtree(TARGET)
    (TARGET / "assets").mkdir(parents=True)

    slides = []
    for tid in TEMPLATE_ORDER:
        src = ROOT / "seed" / "templates" / tid
        meta = json.loads((src / "meta.json").read_text(encoding="utf-8"))
        tpl = json.loads((src / "slides.json").read_text(encoding="utf-8"))
        slide = prepare_slide(tpl["slides"][0], meta["title"])
        slides.append(slide)
        assets = src / "assets"
        if assets.is_dir():
            for f in assets.iterdir():
                if f.is_file():
                    shutil.copy2(f, TARGET / "assets" / f.name)

    seed_meta = {
        "title": "Schlicht",
        "width": 1920,
        "height": 1080,
        "template_order": 0,
        "is_layout_set": True,
        "template_shared": True,
        "default_layout_set": True,
        "logosNotesOrder": [
            "document_title",
            "heading1",
            "heading2",
            "heading3",
            "heading4",
            "heading5",
            "normal",
            "list_item",
            "lighttext",
            "prompt",
            "scripture_block",
            "scripture_inline",
        ],
        "elementZones": {
            "slides": ["document_title", "heading1", "heading2", "heading3", "heading4", "heading5", "normal", "list_item"],
            "footer": ["scripture_ref", "scripture_verse"],
            "custom": ["lighttext", "prompt"],
            "unused": ["meta"],
        },
        "safe_margin": 100,
    }

    (TARGET / "meta.json").write_text(
        json.dumps(seed_meta, ensure_ascii=False, indent=4) + "\n",
        encoding="utf-8",
    )
    (TARGET / "slides.json").write_text(
        json.dumps({"slides": slides}, ensure_ascii=False, indent=4) + "\n",
        encoding="utf-8",
    )
    print(f"Erstellt: seed/layout-sets/{SEED_NAME}/ mit {len(slides)} Layout-Folien.")


if __name__ == "__main__":
    main()
